### PR TITLE
Fix/ewo entry, backtest

### PR DESCRIPTION
**Pull request by zenixls2**
fix: non-handled split between kline prev close -> current open
fix: ewo entry, ewo average price calculation
image

### CODE REVIEW

File: pkg/backtest/matching.go [hunk 1]
```diff
@@ -410,6 +410,15 @@ func (m *SimplePriceMatching) SellToPrice(price fixedpoint.Value) (closedOrders
 func (m *SimplePriceMatching) processKLine(kline types.KLine) {
 	m.CurrentTime = kline.EndTime.Time()
 	m.LastKLine = kline
+	if m.LastPrice.IsZero() {
+		m.LastPrice = kline.Open
+	} else {
+		if m.LastPrice.Compare(kline.Open) > 0 {
+			m.BuyToPrice(kline.Open)
```
Comment (c9s) [inline pkg/backtest/matching.go:417]: should be "sellTo" here? since last price > next open, the price is moving down?

**Comment by zenixls2** [comment]
also fixed bugs in boundary check in chart, and position time vs order update_time issue.

**Comment by zenixls2** [comment]
fix: sqlite3 db lock issue by batch processing (using transaction as a solution)
fix: ewo TP compare direction
feature: add trade report from ewo
image

File: pkg/service/backtest.go [hunk 2]
```diff
@@ -321,8 +320,11 @@ func (s *BacktestService) BatchInsert(kline []types.KLine) error {
 	sql := fmt.Sprintf("INSERT INTO `%s` (`exchange`, `start_time`, `end_time`, `symbol`, `interval`, `open`, `high`, `low`, `close`, `closed`, `volume`, `quote_volume`, `taker_buy_base_volume`, `taker_buy_quote_volume`)"+
 		" values (:exchange, :start_time, :end_time, :symbol, :interval, :open, :high, :low, :close, :closed, :volume, :quote_volume, :taker_buy_base_volume, :taker_buy_quote_volume); ", tableName)
 
-	_, err := s.DB.NamedExec(sql, kline)
-	return err
+	tx := s.DB.MustBegin()
```
Comment (c9s) [inline pkg/service/backtest.go:323]: why do we need transaction here?

**Review by c9s — APPROVED** [review]

File: pkg/types/position.go [hunk 3]
```diff
@@ -122,7 +122,7 @@ func (p *Position) NewProfit(trade Trade, profit, netProfit fixedpoint.Value) Pr
 
 func (p *Position) NewClosePositionOrder(percentage fixedpoint.Value) *SubmitOrder {
 	base := p.GetBase()
-	quantity := base.Mul(percentage)
+	quantity := base.Mul(percentage).Abs()
```
Comment (zenixls2) [inline pkg/types/position.go:125]: @c9s fixed the check here

Comment (c9s) [reply pkg/types/position.go:125]: oh yesssss

**Review by austin362667 — APPROVED** [review]

**Pull request merged**